### PR TITLE
Non-Blocking inserts

### PR DESCRIPTION
**Pull request by graetzer**
# Scope & Purpose

Basic Insert operations now use Futures, and are non-blocking. Additionally we do no longer use the curl based networking backend, but instead a fuerte based one.

- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This change is already covered by existing tests, such as the existing integration tests.

http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/5897/

### CODE REVIEW

File: arangod/Network/ConnectionPool.cpp [hunk 1]
```diff
@@ -153,6 +155,7 @@ void ConnectionPool::pruneConnections() {
         if (list.connections.size() <= _config.maxOpenConnections) {
           break;
         }
+        continue;
```
Comment (jsteemann) [inline arangod/Network/ConnectionPool.cpp]: :+1:

File: arangod/Transaction/Methods.cpp [hunk 2]
```diff
@@ -1281,27 +1287,27 @@ Result transaction::Methods::documentFastPathLocal(std::string const& collection
...
-        }
-        // we found an error number, so let's use it!
-        return Result(num.getNumericValue<int>());
-      }
-    }
-  }
-
-  return Result(defaultErrorCode);
-}
+//static Result resultFromClusterResult(std::shared_ptr<VPackBuilder> const& resultBody,
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp]: This seems to have been moved to Network/Utils. So let's remove the commented-out version here.

File: arangod/Transaction/Methods.cpp [hunk 3]
```diff
@@ -1317,61 +1323,21 @@ OperationResult transaction::Methods::clusterResultDocument(
                                         : TRI_ERROR_ARANGO_CONFLICT),
                              resultBody->steal(), nullptr, OperationOptions{}, errorCounter);
     case rest::ResponseCode::NOT_FOUND:
-      errorCode = TRI_ERROR_ARANGO_DOCUMENT_NOT_FOUND;
+      return network::opResultFromBody(resultBody, TRI_ERROR_NO_ERROR);
       break;
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp]: the `break` statement can now be removed as well
```suggestion
```

File: arangod/Transaction/Methods.cpp [hunk 4]
```diff
@@ -1421,18 +1381,16 @@ OperationResult transaction::Methods::clusterResultRemove(
                              resultBody->steal(), nullptr, options, errorCounter);
     }
     case rest::ResponseCode::BAD:
-      errorCode = TRI_ERROR_INTERNAL;
+      return network::opResultFromBody(resultBody, TRI_ERROR_INTERNAL);
       break;
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp]: the `break` statement can now be removed.
```suggestion
```

File: arangod/Transaction/Methods.cpp [hunk 5]
```diff
@@ -1421,18 +1381,16 @@ OperationResult transaction::Methods::clusterResultRemove(
...
     case rest::ResponseCode::NOT_FOUND:
-      errorCode = TRI_ERROR_ARANGO_DOCUMENT_NOT_FOUND;
+      return network::opResultFromBody(resultBody, TRI_ERROR_ARANGO_DOCUMENT_NOT_FOUND);
       break;
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp]: the `break` statement can now be removed.
```suggestion
```

File: arangod/Transaction/Methods.cpp [hunk 6]
```diff
@@ -3304,23 +3268,23 @@ Result transaction::Methods::resolveId(char const* handle, size_t length,
...
     return res;
   }
-
+  
   // nullptr only happens on controlled shutdown
-  auto cc = arangodb::ClusterComm::instance();
-  if (cc == nullptr) {
-    return res.reset(TRI_ERROR_SHUTTING_DOWN);
-  };
+//  auto cc = arangodb::ClusterComm::instance();
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp]: let's remove all this commented-out code.

File: arangod/Transaction/Methods.cpp [hunk 7]
```diff
@@ -3399,22 +3361,30 @@ Result Methods::replicateOperations(LogicalCollection const& collection,
...
   }
 
-  auto body = std::make_shared<std::string>();
-  *body = payload->slice().toJson();
-
   // Now prepare the requests:
-  std::vector<ClusterCommRequest> requests;
-  requests.reserve(followers->size());
-
+//  std::vector<ClusterCommRequest> requests;
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp]: let's remove this

File: arangod/Transaction/Methods.cpp [hunk 8]
```diff
@@ -3399,22 +3361,30 @@ Result Methods::replicateOperations(LogicalCollection const& collection,
...
+    // TODO we could steal the payload at least once
+    VPackBuffer<uint8_t> buffer;
+    buffer.append(payload->data(), payload->size());
+    
+    network::Headers headers;
+    ClusterTrxMethods::addTransactionHeader(*this, f, headers);
+    auto future = network::sendRequestRetry("server:" + f, requestType,
+                                            path, std::move(buffer), timeout, headers, /*retryNotFound*/true);
+    futures.emplace_back(std::move(future));
+//    auto headers = std::make_unique<std::unordered_map<std::string, std::string>>();
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp]: let's remove this

File: arangod/Transaction/Methods.cpp [hunk 9]
```diff
@@ -3399,22 +3361,30 @@ Result Methods::replicateOperations(LogicalCollection const& collection,
...
-  *body = payload->slice().toJson();
-
   // Now prepare the requests:
-  std::vector<ClusterCommRequest> requests;
-  requests.reserve(followers->size());
-
+//  std::vector<ClusterCommRequest> requests;
+//  requests.reserve(followers->size());
+  
+  std::vector<Future<network::Response>> futures;
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp]: we can consider re-adding `requests.reserve(followers->size());` as a micro optimization here.

File: arangod/Transaction/Methods.cpp [hunk 10]
```diff
@@ -3426,42 +3396,51 @@ Result Methods::replicateOperations(LogicalCollection const& collection,
...
+      }
+      
+      if (!replicationWorked) {
+        Result res = finfo->remove((*followers)[i]);
+        if (res.ok()) {
+          // TODO: what happens if a server is re-added during a transaction ?
+          _state->removeKnownServer((*followers)[i]);
+          LOG_TOPIC("12d8c", WARN, Logger::REPLICATION)
+          << "synchronous replication: dropping follower " << (*followers)[i]
+          << " for shard " << "TODO";//collection.name();
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp]: shall we fix the "TODO" here?

File: arangod/Transaction/Methods.cpp [hunk 11]
```diff
@@ -3426,42 +3396,51 @@ Result Methods::replicateOperations(LogicalCollection const& collection,
...
+        } else {
+          LOG_TOPIC("db473", ERR, Logger::REPLICATION)
+          << "synchronous replication: could not drop follower " << (*followers)[i]
+          << " for shard " << "TODO"/*collection.name()*/ << ": " << res.errorMessage();
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp]: same

File: arangod/Cluster/ClusterMethods.cpp [hunk 12]
```diff
@@ -1199,29 +1224,32 @@ Result createDocumentOnCoordinator(transaction::Methods& trx, std::string const&
...
   if (isManaged && shardMap.size() > 1) {
-    Result res = beginTransactionOnSomeLeaders(*trx.state(), collinfo, shardMap);
+    Result res = beginTransactionOnSomeLeaders(*trx.state(), coll, shardMap);
     if (res.fail()) {
-      return res.errorNumber();
+      return makeFuture(OperationResult(std::move(res)));
     }
   }
+  
+  std::vector<Future<network::Response>> futures;
```
Comment (jsteemann) [inline arangod/Cluster/ClusterMethods.cpp]: shall we add `futures.reserve(shardMap.size());`?

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/7576/

File: 3rdParty/fuerte/src/ConnectionBuilder.cpp [hunk 13]
```diff
@@ -124,43 +123,75 @@ void parseSchema(std::string const& schema,
   }
 }
 
+#define IS_INVALID(c) (c == ' ' || c == '\r' || c == '\n')
+#define LOWER(c)            (unsigned char)(c | 0x20)
+#define IS_ALPHA(c)         (LOWER(c) >= 'a' && LOWER(c) <= 'z')
+#define IS_NUM(c)           ((c) >= '0' && (c) <= '9')
+#define IS_ALPHANUM(c)      (IS_ALPHA(c) || IS_NUM(c))
+#define IS_HOST_CHAR(c)     (IS_ALPHANUM(c) || (c) == '.' || (c) == '-')
+
```
[the remark this answers is not in the document]
Comment (graetzer) [reply 3rdParty/fuerte/src/ConnectionBuilder.cpp]: why bother ?

File: arangod/Cluster/ClusterMethods.cpp [hunk 14]
```diff
@@ -1151,6 +1141,54 @@ int selectivityEstimatesOnCoordinator(std::string const& dbname, std::string con
...
+        tmpBuilder->add(StaticStrings::Error, VPackValue(true));
+        tmpBuilder->add(StaticStrings::ErrorNum, VPackValue(commError));
+        tmpBuilder->close();
+      }
+      resultMap.emplace(sId, std::move(tmpBuilder));
+    } else {
+      std::vector<VPackSlice> const& slices = res.response->slices();
+      auto tmpBuilder = std::make_shared<VPackBuilder>();
+      if (!slices.empty()) {
+        tmpBuilder->add(slices[0]);
```
[the remark this answers is not in the document]
Comment (graetzer) [reply arangod/Cluster/ClusterMethods.cpp]: no

File: arangod/Cluster/ClusterTrxMethods.h [hunk 15]
```diff
@@ -34,41 +34,38 @@ namespace arangodb {
 struct OperationOptions;
 class TransactionState;
 
-class ClusterTrxMethods {
- public:
-  // wrapper Class for static functions.
-  // Cannot be instanciated.
-  ClusterTrxMethods() = delete;
-  ~ClusterTrxMethods() = delete;
+namespace ClusterTrxMethods {
```
[the remark this answers is not in the document]
Comment (graetzer) [reply arangod/Cluster/ClusterTrxMethods.h]: yes

File: arangod/Network/NetworkFeature.cpp [hunk 16]
```diff
@@ -32,15 +32,14 @@
...
-void queueGarbageCollection(std::shared_ptr<arangodb::NetworkFeature> feature,
-                            std::mutex& mutex, arangodb::Scheduler::WorkHandle& workItem,
+void queueGarbageCollection(std::mutex& mutex, arangodb::Scheduler::WorkHandle& workItem,
                             std::function<void(bool)>& gcfunc, std::chrono::seconds offset) {
   bool queued = false;
   {
     std::lock_guard<std::mutex> guard(mutex);
     std::tie(queued, workItem) =
         arangodb::basics::function_utils::retryUntilTimeout<arangodb::Scheduler::WorkHandle>(
-            [feature, &gcfunc, offset]() -> std::pair<bool, arangodb::Scheduler::WorkHandle> {
```
[the remark this answers is not in the document]
Comment (graetzer) [reply arangod/Network/NetworkFeature.cpp]: the shutdown order should prevent this otherwise we are SOL

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/5932/

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/5942/

**Pull request merged**